versionedController/v1/configuration/configuration.go

```diff
@@ -53,79 +59,72 @@ func (configurationController *Controller) Put(c *gin.Context) {
-			return
-		}
-
-		secret.StringData = data
-		_, err = k8s.ClientSet.CoreV1().Secrets(types.DefaultNamespace).Update(c.Request.Context(), secret, metav1.UpdateOptions{})
-		if err != nil {
-			c.String(http.StatusInternalServerError, "Error storing credentials : %v", err)
-			log.Errorln("Error updating secret ", err)
-			return
-		}
-
-		controller.Server.GithubConfig.ClientID = *configurationModel.ClientID
-		controller.Server.GithubConfig.ClientSecret = *configurationModel.ClientSecret
+// updateToK8s updates the value of Oauth secrets in secrets & configmaps
+// TODO: Need to lock this operation, so only one request can update it at a time
+// TODO(vharsh): Use pflags to configure all of these
+func (configurationController *Controller) updateToK8s(c *gin.Context, requestModel *Model) {
+	// 1. Get current configuration from configmap
+	// 2. See if any update is required
+	// 3. Respond back with updated *Model
+	cm, err := k8s.ClientSet.CoreV1().ConfigMaps(types.DefaultNamespace).
+		Get(c.Request.Context(), types.DefaultConfigMap, metav1.GetOptions{})
+	if err != nil {
+		c.JSON(http.StatusPreconditionFailed, gin.H{
+			"message": "Unable to persist config change",
+		})
 	}
-	if configurationModel.EnableGithub != nil {
-		cm, err := k8s.ClientSet.CoreV1().ConfigMaps(types.DefaultNamespace).Get(c.Request.Context(), types.DefaultConfigMap, metav1.GetOptions{})
-		if err != nil {
-			c.String(http.StatusInternalServerError, "Error enabling github login : %v", err)
-			log.Errorln("Error getting configmap ", err)
-			return
-		}
-
-		cm.Data[types.DISABLE_GITHUBAUTH] = strconv.FormatBool(!*configurationModel.EnableGithub)
-		_, err = k8s.ClientSet.CoreV1().ConfigMaps(types.DefaultNamespace).Update(c.Request.Context(), cm, metav1.UpdateOptions{})
-		if err != nil {
-			c.String(http.StatusInternalServerError, "Error enabling github login : %v", err)
-			log.Errorln("Error updating configmap ", err)
-			return
-		}
-
-		controller.Server.Config.DisableGithubAuth = !*configurationModel.EnableGithub
+	githubEnable, _ := strconv.ParseBool(cm.Data[types.DISABLE_GITHUBAUTH])
+	githubEnable = !githubEnable
+	googEnable, _ := strconv.ParseBool(cm.Data[types.DISABLE_GOOGLEAUTH])
+	googEnable = !googEnable
+	githubClientID := cm.Data[types.GITHUB_CLIENT_ID]
+	githubClientSecret := cm.Data[types.GITHUB_CLIENT_SECRET]
+	googClientID := cm.Data[types.GOOGLE_CLIENT_ID]
+	googClientSecret := cm.Data[types.GOOGLE_CLIENT_SECRET]
+	cfgMapModel := Model{
+		GithubClientID:     &githubClientID,
+		GithubClientSecret: &githubClientSecret,
+		EnableGithub:       &githubEnable,
+		GoogleClientID:     &googClientID,
+		GoogleClientSecret: &googClientSecret,
+		EnableGoogle:       &googEnable,
 	}
-
-	c.String(http.StatusOK, "Configured successfully")
+	// update the configmap model with data from the request-model
+	mergo.Merge(&cfgMapModel, &requestModel)
+	// merge the request data
+	mergo.Merge(&cm.Data, &cfgMapModel)
```

Comment:
Why do we need to merge it 2 times. Can't we directly do `mergo.merge(&cm.Data, &requestModel)`

Reply:
One is merging the data in the configmap so that it can be reapplied and one is merging the data in the cfgMapModel so that it can be returned to the client(UI) so that they'd have the updated data. The UI need should not have the full configmap data returned.